packages/core/graph/src/AdjacencyList.js

```diff
@@ -1198,6 +1464,77 @@ export class EdgeTypeMap<TEdgeType> extends SharedTypeMap<
   }
 }
 
+/**
+ * Links a node to another node with an edge of the given type.
+ *
+ * Returns:
+ * - `true` if the edge was added
+ * - `false` if the edge already exists
+ * - `'resizeEdges'` if the edge map is overloaded
+ * - `'reclaimDeletes'` if the edge map has too many deleted edges
+ * - `'resizeNodes'` if the node map is overloaded
+ */
+function link<TEdgeType: number>(
+  from: NodeId,
+  to: NodeId,
+  type: TEdgeType | NullEdgeType,
+  edges: EdgeTypeMap<TEdgeType | NullEdgeType>,
+  nodes: NodeTypeMap<TEdgeType | NullEdgeType>,
+  unloadFactor: number = DEFAULT_PARAMS.unloadFactor,
+): boolean | 'resizeEdges' | 'reclaimDeletes' | 'resizeNodes' {
```

Comment:
wonder if returning multiple different types from this function is optimal. would it be better to create an enum of numbers for these?

Reply:
Good call, I refactored into an enum, and also changed the wording to be more about why the link failed than about what to do about a failure (e.g., `EdgesOverloaded` instead of `'resizeEdges'`)